article/article.md

```diff
@@ -0,0 +1,153 @@
+
+To figure out how best to do this, I spent some time examining the public APIs and interals of both `TileLayer` and `GridLayer` (which the former extends). `GridLayer` maintains a cache of tiles, indexed by `x`, `y`, and `z` (zoom level) coordinates. Based on the user's current view and some configuration settings, `GridLayer` is responsible for deciding when to add or remove tiles, and it controls how these tiles are positioned on the map. The actual tile creation is delegated to a method, `createTile`, that is intended to be overridden in derived classes. Our custom component would need to fetch and render its tiles in `createTile`.
+
+For optimal performance in OpenGL, you typically want to render as much as possible in a single pass or "draw call". I hoped to render multiple tiles per draw call to take best advantage of the hardware, but the design of `GridLayer` necessitated drawing one tile at a time, because `createTile` is called as needed for _each_ tile that must be rendered. There was another issue as well: WebGL rendering is tied to a `canvas` HTML element. So as not to overtax the hardware resources and to prevent costly state changes in the WebGL system, it's best to use just one `canvas` for rendering. Given the design of `GridLayer`, though, which places an HTML element on the DOM _for each tile_, a single `canvas` clearly wouldn't suffice.
+
+Creating a new WebGL context for each new `canvas` added to the DOM seemed too expensive, so I opted instead to create a single offscreen `canvas` used exclusively for WebGL rendering. Each tile would be rendered to this offscreen `canvas` on demand, then the rendered pixels would be copied to a new `canvas` that `GridLayer` would attach to the map. It's not an ideal solution, because when many tiles need to be rendered, many pixels must be copied from the rendering `canvas` to the onscreen `canvas`es, but it seemed the best option available given the design of `GridLayer`.
+
+## Flourishes
+
+By now, the main work was complete, but there were a few more nice-to-have features I was keen to add. First, I wanted to make the floating-point value of the pixel under the cursor available to user-defined mouse events. Since the parent class, `GridLayer`, maintains a cache of the tiles visible onscreen, doing so was pretty easy. I could use the cursor's screen position to determine (1) which tile was under the cursor and (2) the cursor's pixel coordinates within the tile. From there, the float value could be obtained from the binary pixel data with a JavaScript `DataView`:
```

Comment:
I would add a GIF of your tooltips too. Maybe even with a network tab showing no network requests happening!

Reply:
Yeah, I really wanted to do that; just couldn't get it to work with the program (Giphy) I used to create the other animated GIF.

Reply:
Incorporated the screen capture you made - thanks very much!